### PR TITLE
Feat: katex render more equations correctly by splitting delimiters into strict and loose types

### PR DESCRIPTION
**Pull request by yangqian**
# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) to discuss your idea/fix with the community before creating a pull request, and describe your changes before submitting a pull request.

This is to ensure large feature PRs are discussed with the community first, before starting work on it. If the community does not want this feature or it is not relevant for Open WebUI as a project, it can be identified in the discussion before working on the feature and submitting the PR.

**Before submitting, make sure you've checked the following:**

- [X] **Target branch:** Verify that the pull request targets the `dev` branch. **PRs targeting `main` will be immediately closed.**
- [X] **Description:** Provide a concise description of the changes made in this pull request down below.
- [X] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [X] **Testing:** Perform manual tests to **verify the implemented fix/feature works as intended AND does not break any other functionality**. Include reproducible steps to demonstrate the issue before the fix. Test edge cases (URL encoding, HTML entities, types). Take this as an opportunity to **make screenshots of the feature/fix and include them in the PR description**.
- [X] **Agentic AI Code:** Confirm this Pull Request is **not written by any AI Agent** or has at least **gone through additional human review AND manual testing**. If any AI Agent is the co-author of this PR, it may lead to immediate closure of the PR.
- [X] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [X] **Design & Architecture:** Prefer smart defaults over adding new settings; use local state for ephemeral UI logic. Open a Discussion for major architectural or UX changes.
- [X] **Git Hygiene:** Keep PRs atomic (one logical change). Clean up commits and rebase on `dev` to ensure no unrelated commits (e.g. from `main`) are included. Push updates to the existing PR branch instead of closing and reopening.
- [X] **Title Prefix:** To clearly categorize this pull request, prefix the pull request title using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

seprate the latex delimiters into two groups such that many latex equations could be rendered while still being conservative with the commonly used dollar signs.

### Added

- Split KaTeX delimiters into "strict" (\(, \), \[, \], \ce{}, \pu{}, \begin{equation}) and "loose" ($, $$) groups. Strict delimiters are unambiguous LaTeX syntax and now match without requiring surrounding-character checks.

### Fixed

- Prevented inline KaTeX regex from matching across newlines by changing content patterns from [^] to [^\n], fixing cross-line backtracking that could cause intermediate \(...\) expressions to be consumed by a single greedy $...$ match.

---

### Additional Information

Fix issue #12036  and many other latex/katex equation related discussions.
### Screenshots or Videos

Render
```
and \(10^3\)-\(10^4\) and
 \(10^3\) ——
$10^3$——
 \(\gamma\)and
x——\(10^4\)
 \(x\)—— and \(y\) 
 \(\gamma\)and\(\gamma_c\) 
 \(\gamma\) and\(\gamma_c\) 
d $\cdotp$ b
d $\cdot$ b

The Handshaking Lemma states that $\sum_{v \in V} \deg(v) = 2|E|$.
If all 5 vertices have degree 3, then:

$$\sum_{v \in V} \deg(v) = 15$$

which is odd — contradicting the lemma. QED.
```
Before:
image
After:
image
Note that the second dollar sign break is intentional.

render:
```
\(\xi = (\xi_1, \dots, \xi_N)\)xx
\(a^2 + b^2 = c^2\)\(e^{i\pi} + 1 = 0\)
```
Before:
image

After:
image

### Contributor License Agreement

By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](https://github.com/open-webui/open-webui/blob/main/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

> [!NOTE]
> Deleting the CLA section will lead to immediate closure of your PR and it will not be merged in.

### CODE REVIEW

**Comment by yangqian** [comment]
✅ I have personally tested ALL changes in this PR
✅ The production server is running the 0.8.8 latest version. The test server implemented the change. I compared the changes by editing the prompt in both servers.
✅ See before and after screenshots in production and test server.

File: src/lib/utils/marked/katex-extension.ts [hunk 1]
```diff
@@ -76,7 +109,7 @@ function katexStart(src, displayMode: boolean) {
 		let startIndex = -1;
 		let startDelimiter = '';
 		let endDelimiter = '';
-		for (const delimiter of DELIMITER_LIST) {
+		for (const delimiter of ALL_DELIMITERS) {
 			if (delimiter.display !== displayMode) {
 				continue;
 			}
```
[the remark this answers is not in the document]
Comment (yangqian) [reply src/lib/utils/marked/katex-extension.ts]: This is just a rename of variable to reflect the division of delimiters

File: src/lib/utils/marked/katex-extension.ts [hunk 2]
```diff
@@ -95,12 +128,14 @@ function katexStart(src, displayMode: boolean) {
...
 
-		// Check if the delimiter is preceded by a special character.
-		// If it does, then it's potentially a math formula.
-		const f =
+		// Strict delimiters skip the surrounding-char check; loose ones require it
+		const isStrict = !LOOSE_LEFT_SET.has(startDelimiter);
+		const surroundingOk =
+			isStrict ||
 			index === 0 ||
 			indexSrc.charAt(index - 1).match(new RegExp(`[${ALLOWED_SURROUNDING_CHARS}]`, 'u'));
```
[the remark this answers is not in the document]
Comment (yangqian) [reply src/lib/utils/marked/katex-extension.ts]: I tried to keep the change minimum.

**Comment by yangqian** [comment]
The problem still persists in 0.8.9 even though the original test was done in 0.8.8

**Comment by yangqian** [comment]
The problem still persists in 0.8.10 even though the original test was done in 0.8.8

**Comment by danilowanner** [comment]
I can help do a performance test on Safari once this is merged, if helpful.

File: src/lib/utils/marked/katex-extension.ts [hunk 3]
```diff
@@ -1,14 +1,24 @@
-const DELIMITER_LIST = [
+// "Loose" delimiters: $ and $$ are ambiguous (could be currency), so they
+// require surrounding-character checks to avoid false positives.
+const LOOSE_DELIMITERS = [
```
Comment (2000090079) [inline src/lib/utils/marked/katex-extension.ts:3]: Good separation of loose vs strict delimiters. However `LOOSE_LEFT_SET` 
(used later in katexStart) should be auto-derived from this array:

const LOOSE_LEFT_SET = new Set(LOOSE_DELIMITERS.map(d => d.left));

This prevents them drifting out of sync if new loose delimiters 
are added in future.

File: src/lib/utils/marked/katex-extension.ts [hunk 4]
```diff
@@ -27,44 +37,64 @@ const ALLOWED_SURROUNDING_CHARS_REGEX = new RegExp(`[${ALLOWED_SURROUNDING_CHARS
...
 		const { left, right, display } = delimiter;
-		// Ensure regex-safe delimiters
 		const escapedLeft = escapeRegex(left);
 		const escapedRight = escapeRegex(right);
 
 		if (!display) {
-			// For inline delimiters, we match everything
-			inlinePatterns.push(`${escapedLeft}((?:\\\\[^]|[^\\\\])+?)${escapedRight}`);
+			// Inline-only delimiters must not span newlines to prevent greedy cross-line matching
+			inlinePatterns.push(`${escapedLeft}((?:\\\\[^\\n]|[^\\\\\\n])+?)${escapedRight}`);
```
Comment (2000090079) [inline src/lib/utils/marked/katex-extension.ts:56]: The fix from `[^]` to `[^\n]` correctly prevents inline delimiters 
from matching across newlines — good fix for the greedy cross-line 
backtracking bug.

One question: should the same `[^\n]` constraint also apply to 
block delimiters that double as inline? Line below still uses `[^]`:

`inlinePatterns.push(\`${escapedLeft}(?!\n)((?:\\[^]|[^\\])+?)(?!\n)${escapedRight}\`)`

That pattern could still greedily match across lines for `\[...\]` 
when used inline.

File: src/lib/utils/marked/katex-extension.ts [hunk 5]
```diff
@@ -94,9 +125,8 @@ function katexStart(src, displayMode: boolean) {
 				// Inline: \( or \ce{ or \pu{
 				if (next !== '(' && next !== 'c' && next !== 'p') continue;
 			}
-			if (i === 0 || ALLOWED_SURROUNDING_CHARS_REGEX.test(src.charAt(i - 1))) {
-				return i;
-			}
+			// \ delimiters are strict — no surrounding-char check needed
```
Comment (2000090079) [inline src/lib/utils/marked/katex-extension.ts:128]: Removing the surrounding-char check for strict `\` delimiters is the 
correct approach. However, consider adding an early return comment or 
guard for the case where `src` starts at position 0 with a strict 
delimiter — currently that path falls through correctly, but it's 
not immediately obvious to future readers why there is no `i === 0` 
guard here unlike the `$` branch above.

A short inline comment like:
// strict delimiters are always valid regardless of preceding char
would help maintainability.
